### ISSUE TITLE
realpath not found on macOS 12.2

### ISSUE THREAD

**Issue by vikingtoby** [sshuttle/sshuttle#722]
Running `sshuttle --sudoers` or `sshuttle --sudoers --sudoers-user <username>` throws the following error after updating to 1.1.0 using Homebrew, worked prior to updating.
Tried complete reinstall as well, in case some legacy stuff was left from previous version, same thing though.

```
$ sshuttle --sudoers
# WARNING: When you allow a user to run sshuttle as root,
# they can then use sshuttle's --ssh-cmd option to run any
# command as root.
Password:
/usr/local/bin/sudoers-add: line 54: realpath: command not found
Invalid sudoers filename: Final sudoers file location () does not begin with /etc/sudoers.d
Failed updating sudoers file.
```

**Comment by vikingtoby** [comment]
Reverting the changes in commit: https://github.com/sshuttle/sshuttle/commit/58c264ff1c28ca2f780c86ba79ba58cd21050617
Seems to fix it, so "realpath" is not included as a requirement?

(Not much of a Python guy).

**Comment by brianmay** [comment]
realpath is not Python, so sshuttle itself cannot specify it as a requirement.

Most Linux distributions include this as part of core installation, I guess OSX doesn't.

```
brian@canidae:~$ dpkg -S /usr/bin/realpath
coreutils: /usr/bin/realpath
brian@canidae:~$ apt-cache show coreutils
Package: coreutils
Source: coreutils (8.32-4)
Version: 8.32-4+b1
Essential: yes
Installed-Size: 17478
Maintainer: Michael Stone <mstone@debian.org>
Architecture: amd64
Pre-Depends: libacl1 (>= 2.2.23), libattr1 (>= 1:2.4.44), libc6 (>= 2.28), libgmp10, libselinux1 (>= 3.1~)
Description-en: GNU core utilities
 This package contains the basic file, shell and text manipulation
 utilities which are expected to exist on every operating system.
 .
 Specifically, this package includes:
 arch base64 basename cat chcon chgrp chmod chown chroot cksum comm cp
 csplit cut date dd df dir dircolors dirname du echo env expand expr
 factor false flock fmt fold groups head hostid id install join link ln
 logname ls md5sum mkdir mkfifo mknod mktemp mv nice nl nohup nproc numfmt
 od paste pathchk pinky pr printenv printf ptx pwd readlink realpath rm
 rmdir runcon sha*sum seq shred sleep sort split stat stty sum sync tac
 tail tee test timeout touch tr true truncate tsort tty uname unexpand
 uniq unlink users vdir wc who whoami yes
Description-md5: d0d975dec3625409d24be1238cede238
Multi-Arch: foreign
Homepage: http://gnu.org/software/coreutils
Tag: admin::configuring, implemented-in::c, interface::commandline,
 role::program, scope::utility, suite::gnu, works-with::file
Section: utils
Priority: required
Filename: pool/main/c/coreutils/coreutils_8.32-4+b1_amd64.deb
Size: 2866720
MD5sum: c4c2510ccf8e77a536e3a5d55f13e07b
SHA256: 3558a412ab51eee4b60641327cb145bb91415f127769823b68f9335585b308d4
```

I think this might need to be fixed in the Homebrew packaging.

Having said that it might be possible to rewrite the test not to require realpath. It was added by https://github.com/sshuttle/sshuttle/commit/8c5ffc9e729aeedf25a6c6c26e62406b0c58e436.

**Comment by skuhl** [comment]
I believe the realpath binary is provided by the "coreutils" package in brew. Does installing that allow sshuttle to find realpath successfully? If so, homebrew might want to make sshuttle depend on coreutils.

**Comment by brianmay** [comment]
This is not fixed by #727.

**Comment by jBouyoud** [comment]
I've the same kind of error on MacOS 11.2, sshuttle 1.1.0
```
Invalid sudoers filename: Final sudoers file location (/private/etc/sudoers.d/sshuttle_auto) does not begin with /etc/sudoers.d
Failed updating sudoers file.
```

**Comment by Steuard** [comment]
I'm pretty sure that for anyone installing sshuttle via a package manager, this is just a dependency that needs to be added in that package management system. Homebrew needs to list a coreutils dependency, and MacPorts needs to list a realpath dependency ("port install realpath" fixes the problem). Maybe the installation instructions for sshuttle ought to mention the need for that utility, but otherwise I don't think it's something that the sshuttle code itself can or should fix.

**Comment by beebird** [comment]
It's not caused by missing of realpath: https://github.com/sshuttle/sshuttle/commit/58c264ff1c28ca2f780c86ba79ba58cd21050617#commitcomment-67943257

**Comment by brianmay** [comment]
On the contrary, the message looks definite to me:

```
realpath: command not found
```

So maybe this is not the only problem but it definitely does look like a problem to me.

**Comment by Steuard** [comment]
It's both: the realpath utility isn't present by default on macOS, and also the true macOS path to /etc/sudoers.d is different than on other systems. (Speaking here as someone who had to fix both parts of the issue to get things working.)

(Actually, I had one _further_ issue before my configuration actually worked, which I should submit separately but may not have time to right now. The specific command permitted in the sudoers file needed to start with the actual invocation of python, not with /usr/bin/env PYTHONPATH=... .)

**Comment by skuhl** [comment]
> It's both: the realpath utility isn't present by default on macOS,

If you used a packaging system to package sshuttle, this issue is actually an issue of not installing the correct dependencies. However, I also don't think sshuttle was clear that this was an additional requirement when it was added. There was probably an assumption that the executable was available everywhere since it is present on most/all(?) linux machines.. This can be manually fixed as described here: https://github.com/sshuttle/sshuttle/issues/722#issuecomment-1029393946

> and also the true macOS path to /etc/sudoers.d is different than on other systems. (Speaking here as someone who had to fix both parts of the issue to get things working.)

There was a pull request at one potential solution to this, and also discussion about potentially just removing this whole feature entirely. See this pull request: https://github.com/sshuttle/sshuttle/pull/732

One workaround for this problem is to simply use --sudoers-no-modify and manually copy/add/apply the sudo rules in the proper location. The --sudoers option potentially misleads people into thinking that it allows only the sshuttle command to be run as root without a password, when a simple trick can be employed to make it run any command as root without a password.

> (Actually, I had one _further_ issue before my configuration actually worked, which I should submit separately but may not have time to right now. The specific command permitted in the sudoers file needed to start with the actual invocation of python, not with /usr/bin/env PYTHONPATH=... .)

This should be fixed in git: https://github.com/sshuttle/sshuttle/pull/727

**Comment by brianmay** [comment]
One of the goals of shuttle has always been to have minimal dependencies. It makes me easy that we need to have a new dependency on realpath just for one feature that is currently somewhat controversial. Still, if we want to keep this code, not sure there is much alternative here.

**Comment by skuhl** [comment]
I'm preparing a pull request that removes --sudoers but keeps the --sudoers-no-modify option. This would remove the realpath dependency and simplify things to some degree. I hope to submit it in a day or two once I have a chance to take a look through it again.

**Comment by beebird** [comment]
> I'm preparing a pull request that removes --sudoers but keeps the --sudoers-no-modify option. This would remove the realpath dependency and simplify things to some degree. I hope to submit it in a day or two once I have a chance to take a look through it again.

Hi, I'm afraid just removing `--sudoers` doesn't resolve the problem... I have tried manually create sudoers file:

`sshuttle --sudoers-no-modify | sudo tee  /etc/sudoers.d/sshuttle_auto
`
but I still encountered the following error:

> sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
> sudo: a password is required
> env: doas: No such file or directory
> fw: fatal: You must be root (or enable su/sudo) to set the firewall
> c : fatal: All attempts to elevate privileges failed.

It's weird.  Can you take a look? @skuhl

**Comment by beebird** [comment]
It seems like I found the reason:
in previous version,  `/usr/bin/env` was appended after `sudo` in the concatenated string, but recent commits changed the behavior.

logs from old version:
`/usr/bin/sudo -p [local sudo] Password:  /usr/bin/env PYTHONPATH=/xxxx/sshuttle/1.0.5_1/libexec/lib/python3.10/site-packages /xxx/sshuttle/1.0.5_1/libexec/bin/python3.10 /xxx/bin/sshuttle -v -v --method auto --firewall
`

logs from current version:
`c : Starting firewall manager with command: ['/usr/bin/env', 'PYTHONPATH=/xxxx/sshuttle/1.1.0/libexec/lib/python3.10/site-packages', '/usr/bin/sudo', '-p', '[local sudo] Password: ', '/xxxx/sshuttle/1.1.0/libexec/bin/python3.10', '/xxxxx/bin/sshuttle', '-v', '-v', '--method', 'auto', '--firewall']
sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
sudo: a password is required`

**Comment by vishvish** [comment]
@beebird your fix of

```
sshuttle --sudoers-no-modify | sudo tee /etc/sudoers.d/sshuttle_auto
```

Worked for me with MacOS 11.6.4 (20G417) and in iTerm2 with Tmux.

**Comment by skuhl** [comment]
@beebird Pull request #727 has already corrected this ordering problem in git.

**Comment by skuhl** [comment]
The original problem mentioned in this ticket has been fixed because we have removed use of realpath in #745.

The other issues mentioned here have also been fixed in git:  The ordering problem of commands was fixed in #727.  Also, the error message involving a message about /private/etc/... when using --sudoers is also resolved by #745 because that feature is removed entirely.

Although there isn't a release containing these fixes, I believe they are fixed in git and this issue can be closed.

**Comment by beebird** [comment]
Well, then it's weird...   Finally I find a workaround by setting ` /etc/sudoers.d/sshuttle_auto` as follows which is consistent with the order listed in the error log above:
```diff
-  Cmnd_Alias SSHUTTLE7CA = /usr/bin/env PYTHONPATH=/opt/homebrew/Cellar/sshuttle/1.1.0/libexec/lib/python3.10/site-packages /opt/homebrew/Cellar/sshuttle/1.1.0/libexec/bin/python3.10 /opt/homebrew/bin/sshuttle *

+  Cmnd_Alias SSHUTTLE7CA = /opt/homebrew/Cellar/sshuttle/1.1.0/libexec/bin/python3.10 /opt/homebrew/bin/sshuttle *
````
Luckily it works (although there's no `PYTHONPATH` in my sudoer file)

**Comment by skuhl** [comment]
Are you using sshuttle from git? Probably best to move the discussion to issue #723 where this particular problem was originally reported.  I just checked it again with sshuttle from git on a MacOS machine and don't see any remaining problems.

**Comment by beebird** [comment]
Oh my bad ,didn't notice that #727 hasn't been included in release 1.1.0 yet.  Nope I'm using MacOS brew version, will wait for next release, thanks!

**Comment by brianmay** [comment]
Since 5719d424de732d218ade4670fd0d987b8aad2736 was merged, this error is no longer applicable.
